Add new Pop-up implementation

luqasn opened the pull request:
Aims to fix https://github.com/TheBlob42/idea-which-key/issues/97 (customization not fully implemented yet).

It is rebuilt using an embedded code editor instead of a html view in order to "blend in" better with the content you are working on. The original idea was to align the characters in the popup pixel-perfect with the characters of the main code editor, but I got lazy :D

I extracted an interface for showing the popup so that users can choose the "traditional" popup or this new one.

The work isn't done yet, but I wanted to get some feedback on how it works for other people and if they like it.

- [x] "icon" support (nerd fonts)
- [x] modifiers rendered as icons instead of text 
- [x] different styles: bottom or right side
- [ ] color support per item
- [ ] allow to configure default colors (settings dialog?/config)
- [x] allow configuring layout (settings dialog?/config)
- [ ] add prefix info rendering and ESC hint at the bottom
- [ ] add new style for "bottom of ide" (out of editor), more or less like current popup
- [ ] come up with a solution for popup that becomes too large to fit on screen. pagination like which-key.nvim and the ideavim bottom panel?


This is how it looks on my machine:

bottom positioning:
![image](https://github.com/user-attachments/assets/0c75810e-19d7-4240-aef6-7636da3e3d0e)

right side positioning (single columns):
![image](https://github.com/user-attachments/assets/c5c8a027-23cd-441c-97a5-f310920f20e4)

luqasn committed:
Refactor popup code into interface

in order to be able to provider alternative popup implementations

luqasn committed:
Add new popup provider

luqasn committed:
WIP support for different placements

TheBlob42 commented:
Sorry for the late response, I currently don't find the time to look into this after work :disappointed: 

Your work looks really promising (only checked the screenshots, not the code yet), I hope I can check the code soon to give you some detailed feedback if possible

xsen commented:
Is there any chance this PR will be completed and merged to main?

luqasn commented:
Hi, sure! I started rebasing it, but then got frustrated/distracted.
I see some interest in it though, so I'll have another stab at it in the coming weeks.

xsen commented:
Thank you so much for your time and effort on this! Really looking forward to it! 🙏

luqasn committed:
wip

luqasn commented:
I have rebased my branch and added very basic configuration options. Feel free to try it!

I have uploaded a build here, install it "manually" inside intellij: https://github.com/luqasn/idea-which-key/releases/tag/v0.12.0-preview1

Then configure the new popup:

```
let g:WhichKey_PopupType="new"
let g:WhichKey_PopupStyle="bottom"
" or:
let g:WhichKey_PopupStyle="right"
```

Let me know what you think!

xsen commented:
I've tried it, and it looks great! I've added a few icons for myself, and I like it.

It seems the font size setting let g:WhichKey_FontSize = 30 is not being applied.

Also, it would be wonderful to have the ability to set margins/padding in the "bottom" mode.

<img width="2013" height="1428" alt="image" src="https://github.com/user-attachments/assets/54501494-3112-4b19-8729-287756513584" />

luqasn commented:
Thanks for trying it out!

It was part of the "design decisions" to use the font settings of the main editor for the popup so that it feels more like a part of the editor (like in neovim). But I acknowledge that for your font size, it is a bit impractical :sweat_smile: 
And I don't see why it shouldn't use a different font size if the user has set one.

Will have a stab at it later. Please keep testing and providing feedback!

shaunek-hero commented:
I just tried the preview release and I absolutely love the right side popup style, thanks guys for working on this. And I also like how it looks a bit more native in this preview. I have really loved this plugin since I found it last year. Preesh!

Regarding the topic of styling the stuff inside the popup, a tiny bit more padding/margin between the border and the text would be nice, it is kinda cramped looking now. I have pretty small font size as well and I'm unsure if I would want to adjust the font size, but if you do that setting I'll probably try it out!

luqasn committed:
Make some stuff configurable

luqasn commented:
I have added some configuration settings now: https://github.com/luqasn/idea-which-key/releases/download/v0.12.0-preview2/IDEA.Which-Key-0.12.0-preview2.zip

It respects the existing `WhichKey_FontSize` setting and also has following new settings:
```
WhichKey_Padding
WhichKey_PaddingHorizontal
WhichKey_PaddingVertical
WhichKey_Margin
WhichKey_MarginHorizontal
WhichKey_MarginVertical
WhichKey_Spacing
```
All of these are measured in "characters", so the size of a character in the editor window that you open the popup in.
For some reason the margin "feels" like it is not really the width of a character, but not sure if that's a bug in the code or in my brain or eyes.

xsen commented:
Thanks for the new version and the configuration options!

I've tested it and found a possible issue with the padding:

With this configuration:
```vim
let g:WhichKey_PaddingHorizontal = 2
let g:WhichKey_PaddingVertical = 2
```
The vertical padding appears larger (or taller) than the horizontal one.
<img width="1610" height="247" alt="image" src="https://github.com/user-attachments/assets/32c788f4-0d1c-4440-92d0-27572601864b" />



This effect becomes even more pronounced when I change the font size:
```vim
let g:WhichKey_FontSize = 16
let g:WhichKey_PaddingHorizontal = 2
let g:WhichKey_PaddingVertical = 2
```
<img width="1666" height="477" alt="image" src="https://github.com/user-attachments/assets/3c2410c8-90d3-4d81-980f-964dc0df0fe0" />
The top margin seems to increase significantly with these settings.

However, I found a configuration that looks good to me:
```vim
let g:WhichKey_MarginHorizontal = 10
let g:WhichKey_MarginVertical = 4

let g:WhichKey_PaddingHorizontal = 2
let g:WhichKey_PaddingVertical = 1
```
<img width="1676" height="1428" alt="image" src="https://github.com/user-attachments/assets/d234ec27-e41d-47dc-afbc-67b5bf4eb657" />

luqasn commented:
The difference is expected because all values are measures in "characters" as to emulate how it would render in a terminal – and a character is taller that it is wide. But I must admit that that's maybe not the best idea. 

It would probably look better if the padding was uniform.

I also find that your popup has too much blank space between the columns, but I am not sure how to fix that.
I'll have another look at the original which-key.
